fdb-extensions/src/main/java/com/apple/foundationdb/async/rtree/ByNodeStorageAdapter.java

```diff
@@ -0,0 +1,272 @@
+                        itemSlots = Lists.newArrayListWithExpectedSize(nodeSlotObjects.size());
+                    }
+                    itemSlots.add(ItemSlot.fromKeyAndValue(itemSlotKeyTuple, itemSlotValueTuple, getHilbertValueFunction()));
+                    break;
+
+                case INTERMEDIATE:
+                    final Tuple childSlotKeyTuple = Tuple.fromList(nodeSlotItems.subList(0, ChildSlot.SLOT_KEY_TUPLE_SIZE));
+                    final Tuple childSlotValueTuple = Tuple.fromList(nodeSlotItems.subList(ChildSlot.SLOT_KEY_TUPLE_SIZE, nodeSlotItems.size()));
+
+                    if (childSlots == null) {
+                        childSlots = Lists.newArrayListWithExpectedSize(nodeSlotObjects.size());
+                    }
+                    childSlots.add(ChildSlot.fromKeyAndValue(childSlotKeyTuple, childSlotValueTuple));
+                    break;
+                default:
+                    throw new IllegalStateException("unknown node kind");
+            }
+        }
+
+        Verify.verify((nodeKind == NodeKind.LEAF && itemSlots != null) ||
+                      (nodeKind == NodeKind.INTERMEDIATE && childSlots != null));
+
+        return nodeKind == NodeKind.LEAF
+               ? new LeafNode(nodeId, itemSlots)
+               : new IntermediateNode(nodeId, childSlots);
+    }
+
+    @Nonnull
+    @Override
+    public <S extends NodeSlot, N extends AbstractNode<S, N>> AbstractChangeSet<S, N>
+            newInsertChangeSet(@Nonnull final N node, final int level, @Nonnull final List<S> insertedSlots) {
+        return new InsertChangeSet<>(node, level, insertedSlots);
+    }
+
+    @Nonnull
+    @Override
+    public <S extends NodeSlot, N extends AbstractNode<S, N>> AbstractChangeSet<S, N>
+            newUpdateChangeSet(@Nonnull final N node, final int level,
+                               @Nonnull final S originalSlot, @Nonnull final S updatedSlot) {
+        return new UpdateChangeSet<>(node, level, originalSlot, updatedSlot);
+    }
+
+    @Nonnull
+    @Override
+    public <S extends NodeSlot, N extends AbstractNode<S, N>> AbstractChangeSet<S, N>
+            newDeleteChangeSet(@Nonnull final N node, final int level, @Nonnull final List<S> deletedSlots) {
+        return new DeleteChangeSet<>(node, level, deletedSlots);
+    }
+
+    private class InsertChangeSet<S extends NodeSlot, N extends AbstractNode<S, N>> extends AbstractChangeSet<S, N> {
+        @Nonnull
+        private final List<S> insertedSlots;
+
+        public InsertChangeSet(@Nonnull final N node, final int level, @Nonnull final List<S> insertedSlots) {
+            super(node.getChangeSet(), node, level);
+            this.insertedSlots = ImmutableList.copyOf(insertedSlots);
+        }
+
+        @Override
+        public void apply(@Nonnull final Transaction transaction) {
+            super.apply(transaction);
+            if (getPreviousChangeSet() == null) {
+                persistNode(transaction, getNode());
```

Comment:
By node storage layout only needs to write out the actual node once. It needs to update the node slot index (if we maintain it) in a more sophisticated manner, so that's why I cannot just collapse all the change sets into one.

Reply:
Maybe I should put a comment here.